pgmpy/base/base.py

```diff
@@ -0,0 +1,948 @@
+    [('A', 'B', {'type': '->'}), ('B', 'C', {'type': '->'})]
+
+    **Nodes:**
+
+    Add one node,
+
+    >>> from pgmpy.base import _CoreGraph
+    >>> G = _CoreGraph()
+    >>> G.add_node("A")
+    >>> G.nodes
+    NodeView(('A',))
+
+    **Edges:**
+
+    G can also be grown by adding edges.
+
+    Add one edge,
+
+    >>> from pgmpy.base import _CoreGraph
+    >>> G = _CoreGraph()
+    >>> G.add_edge("A", "B", "->")
+    >>> G.get_edges(type=True)
+    [('A', 'B', {'type': '->'})]
+
+    Remove one edge,
+
+    >>> edges = [("A", "B", "->"), ("B", "C", "->"), ("C", "D", "--")]
+    >>> G = _CoreGraph(ebunch=edges)
+    >>> G.remove_edge("A", "B", "->")
+    >>> G.get_edges(type=True)
+    [('B', 'C', {'type': '->'}), ('C', 'D', {'type': '--'})]
+
+    **Exposures, Outcomes, and Latents:**
+
+    >>> edges = [("A", "B", "->"), ("B", "C", "->"), ("D", "C", "-o")]
+    >>> G = _CoreGraph(ebunch=edges)
+
+    **Roles:**
+
+    Add node's role.
+
+    >>> G.exposures = "A"
+    >>> G.outcomes = "C"
+    >>> G.latents = "D"
+
+    Checks for the node's role.
+
+    >>> G.exposures
+    {'A'}
+    >>> G.outcomes
+    {'C'}
+    >>> G.latents
+    {'D'}
+
+    In addition to 'exposure', 'outcomes', and 'latents', you can add custom roles.
+
+    >>> edges = [("A", "B", "->"), ("B", "C", "->"), ("D", "C", "-o")]
+    >>> G = _CoreGraph(ebunch=edges)
+    >>> G.with_role("Custom_role", "A", inplace=True)
+    >>> G.with_role("latents", "D", inplace=True)
+    >>> G.get_role_dict()
+    {'latents': ['D'], 'Custom_role': ['A']}
+    >>> G.without_role("Custom_role", "A", inplace=True)
+    >>> G.get_role_dict()
+    {'latents': ['D']}
+    """
+
+    SUPPORTED_EDGE_TYPES = set(["--", "-o", "o-", "->", "<-", "o>", "<o", "<>", "oo"])
+
+    def __init__(
+        self,
+        ebunch: Iterable[tuple[Hashable, Hashable, Hashable]] = None,
+        exposures: Optional[set[Hashable]] = None,
+        outcomes: Optional[set[Hashable]] = None,
+        latents: Optional[set[Hashable]] = None,
+        roles=None,
+    ):
+        super().__init__()
+        if ebunch:
+            self.add_edges_from(ebunch)
+
+        self.exposures = set() if exposures is None else set(exposures)
+        self.outcomes = set() if outcomes is None else set(outcomes)
+        self.latents = set() if latents is None else set(latents)
+
+        if roles is None:
+            roles = {}
+        elif not isinstance(roles, dict):
+            raise TypeError("Roles must be provided as a dictionary.")
+
+        # set the roles to the vertices as networkx attributes
+        for role, vars in roles.items():
+            self.with_role(role=role, variables=vars, inplace=True)
+
+    # ----------------------------------------------------------------------
+    # Public API (or Public Methods)
+    # ----------------------------------------------------------------------
+
+    def add_edge(
+        self,
+        u: Hashable,
+        v: Hashable,
+        type: str = None,
```

Comment:
I think we should rename `type` to `edge_type` as `type` is a keyword in Python.